### DISCUSSION TITLE
Restricting access to 2-sided relationships?

### DISCUSSION THREAD

**Discussion by marcuswohlmuth** [keystonejs/keystone#8284] [Questions] [answered]
My team is currently facing a problem with 2-sided relationships and their (lack of) access control.
We have a lot of the typical `parent-children` relationships in our project, so naturally, we would like to restrict access to these children to whoever "owns" the parent.
Basic access control (will refer to ACL from now on) works fine when it comes to restricting reading, creating, updating and deleting the children themselves. However, it seems like there is no proper way of restricting the actual relationship from the parent's side. Because essays can be confusing, let me demonstrate with an example based on the project starter example (Posts, Authors), and lets say we already configured ACL on those lists the way I described above:
- `Post A` is a child of Author (User) `User 1` and can only be edited by `User 1`
- `Post B` is a child of Author `User 2` and can only be edited by `User 2`
- No matter how ACL is configured, `User 1` can always update `User 1` and connect `Post B` via the relationship.

Now, I know there are workarounds, as usual, and I will list the ones we thought of below. However, additionally I would like to hear your thoughts on how this can be handled or how other teams might have handled this in their projects already.

Known Workarounds:
1. Disabling Access to the `User.posts` field completely via ACL (read -> false, update -> false, delete -> false). This would make the child only editable (and most importantly connectable) via the child itself, enforcing the ACL on the child to only be updated by the respective owner. However, nested `creates` or `connects` from the Parents side would no longer be possible, rendering the field on the parent and the 2-sided relationship in general almost useless, as a simple relationship on the child would suffice for this.
2. Implementing a `validateInput` Hook on the parent's side, checking if all connected children are indeed "my" children prior to connecting. I don't know how well this works, because of how KeystoneJS processes nested creates and connects.

**Comment by borisno2** [comment, marked as answer]
Hi @marcuswohlmuth 
Thanks for starting the discussion. 

Field level access control on the `User.posts` relationship field is what I would recommend, except you should make this a function that returns `true` or `false`.  For example you could set `create` to to only admins (don't allow a user to be created while connecting/creating a post), but set `update` to allow the nested `create`, OR `connect` to posts that are not connected to an existing user. This could look something like (**Note: please ensure you test any Access Control throughly, and that is works for your particular use case - this is just to be used as an example**):
```ts
    posts: relationship({
      ref: 'Post.author',
      many: true,
      access: {
        create: ({ session }) => session?.data.isAdmin,
        update: ({ inputData, session }) => {
          if (!session) {
            return false;
          }
          if (session.data.isAdmin) {
            return true;
          }
          // Don't allow a user to use connect
          if (inputData.posts?.connect) {
              return false
          }
          // if using disconnect or create then allow
          return true;
        },
      },
    }),
```
If you want to get really want to allow `connect` you can check the posts the user is trying to connect to and allow if the author is `null`. To do this you can make `update` a `async` function and pass in `context`.

See https://keystonejs.com/docs/config/access-control#field-access-control for more information about field access control.
